lib/ecto/changeset.ex

```diff
@@ -506,10 +583,17 @@ defmodule Ecto.Changeset do
   """
   @spec put_change(t, atom, term) :: t
   def put_change(%Changeset{} = changeset, key, value) do
-    update_in changeset.changes, &put_change(changeset.model, &1, key, value)
+    changeset = ensure_has_types(changeset)
+    type = Map.get(changeset.types, key)
+    update_in changeset.changes, &put_change(changeset.model, &1, key, value, type)
+  end
+
+  defp put_change(_model, acc, key, value, {:embed, embed}) do
+    value = Ecto.Embedded.wrap_change(embed, value)
+    Map.put(acc, key, value)
```

Comment:
For regular fields we check whether the new value is equal to the one in the model, and in that case we either skip adding this to the list of changes or remove it from that list. 

> If the change is already present, it is overridden with
> the new value, also, if the change has the same value as
> the model, it is not added to the list of changes.

I was thinking about something similar, but I think its too hard to do - at least now.